### PR TITLE
dialects: riscv: Add basic register allocator strategy

### PR DESCRIPTION
**Pull request by adutilleul**
The current register allocator for the RISC-V dialect exclusively utilizes j-registers for unallocated values. However, @compor and I propose an improvement over the existing allocator by introducing a naive strategy that leverages the available real RISC-V registers for each block. In case the real registers are exhausted, the allocator gracefully falls back to j-registers.

This enhancement brings forth several important questions that need to be addressed at this stage:

- Determining the Set of Available Registers: The choice of available registers depends on the ABI  and the adopted calling convention. So, we probably want to determine a target for now and see how it involves other topics (like function handling see https://github.com/xdslproject/xdsl/pull/929).

- Efficiency Requirements for Future Work: We might need to establish the desired level of efficiency for the register allocator (linear scan, ...) and the relevant use cases.

Moreover, this improvement is contingent on the merging of the LabelOp into the main branch. For further details regarding the LabelOp merge, please refer to the following pull request https://github.com/xdslproject/xdsl/pull/994.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Very cool. Would it make sense to have some filecheck test cases?

**Comment by compor** [comment]
> Very cool. Would it make sense to have some filecheck test cases?

Eventually yes, but IMHO it will be something very brittle at this point due to the way we "hand out" available registers.
So, not really sure.

**Comment by tobiasgrosser** [comment]
> > Very cool. Would it make sense to have some filecheck test cases?
> 
> Eventually yes, but IMHO it will be something very brittle at this point due to the way we "hand out" available registers. So, not really sure.

That's the point. If it is brittle we want to see this exposed.

File: xdsl/transforms/riscv_register_allocation.py [hunk 1]
```diff
@@ -27,14 +90,38 @@ def allocate_registers(self, module: ModuleOp) -> None:
...
 class RISCVRegisterAllocation(ModulePass):
     """
-    Allocates unallocated registers in the module. Currently sets them to an infinite set
-    of `j` registers.
+    Allocates unallocated registers in the module.
     """
 
     name = "riscv-allocate-registers"
 
+    allocation_type: Union[str, None] = None
```
Comment (AntonLydike) [inline xdsl/transforms/riscv_register_allocation.py:110]: I wouls suggest something like:

```py
    allocation_type: str = 'global-jregs'
```
Or whatever the magic constant is.

**Review by AntonLydike — CHANGES_REQUESTED** [review]
Overall a step in a very good direction! Just some minor things I'd like to see cleaned up. If you need help feel free to drop by my office!

File: xdsl/riscv_asm_writer.py [hunk 2]
```diff
@@ -91,6 +91,12 @@ def print_assembly_instruction(op: Operation, output: IO[str]) -> None:
                 desc = f"{op.directive.data}"
             print(desc, file=output)
             return
+        case riscv.LabelOp():
+            desc = f"{op.label.data}:"
+            if op.comment is not None and op.comment.data:
+                desc += f"    # {op.comment.data}"
+            print(desc, file=output)
+            return
```
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/riscv_asm_writer.py:99]: @AntonLydike correct, this is based on #994 (also pending).

File: xdsl/transforms/riscv_register_allocation.py [hunk 3]
```diff
@@ -27,14 +90,38 @@ def allocate_registers(self, module: ModuleOp) -> None:
                     self.idx += 1
 
 
+class RegisterAllocationAlgorithm(Enum):
+    """
+    The register allocation algorithm to use.
+    """
+
+    GlobalJRegs = 0
+    BlockNaive = 1
```
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/transforms/riscv_register_allocation.py:99]: What do you mean by other values? Is this a tooling deficiency in typing or a general thing? I cannot find a lot of support on this? Could you please clarify? I thought the point of an enum was to represent a custom type.
Would using `IntEnum` be better?

**Review by nazavode — COMMENTED** [review]
Beautiful, thanks. No additional comments since I would have added the same remarks @AntonLydike pointed out already (especially having #994 merged first).

File: xdsl/riscv_asm_writer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/riscv_asm_writer.py]: Changed the base of thie PR to exclude label op changes

File: requirements-optional.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ nbval<0.11
 filecheck<0.0.24
 lit<17.0.0
 pre-commit==3.3.2
+git+https://github.com/antonlydike/riscemu.git@25d059da090760862f9143478524ed6daf0ab449#egg=riscemu
```
Comment (superlopuh) [inline requirements-optional.txt:11]: We can't actually include this yet, or at least would need to do this in a separate PR where we can discuss the extra dependency separately. I don't think the register allocation requires this change, even if it's useful for debugging, so I'd recommend taking out all riscemu related changes for now.

File: xdsl/transforms/riscv_register_allocation.py [hunk 5]
```diff
@@ -1,19 +1,80 @@
+from dataclasses import dataclass
 from xdsl.dialects.builtin import ModuleOp
 from xdsl.dialects.riscv import Register, RegisterType, RISCVOp
 from xdsl.ir import MLContext
 from xdsl.passes import ModulePass
 
 
-class RegisterAllocator:
+class RegisterAllocatorStrategy:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:8]:
```suggestion
class AbstractRegisterAllocator(ABC):
```

File: xdsl/transforms/riscv_register_allocation.py [hunk 6]
```diff
@@ -1,19 +1,80 @@
...
 
     def __init__(self) -> None:
         self.idx = 0
 
+        """
+        Since we've got neither right now a handling of a consistent ABI nor of a calling convention,
+        let's just assume that we have all the registers available for our use except the one explicitly reserved by the default riscv ABI.
+        """
+
+        self.available_registers = list(Register.ABI_INDEX_BY_NAME.keys())
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:35]: I'd recommend using sets not lists

**Review by superlopuh — CHANGES_REQUESTED** [review]
Thank you! Would be good to iterate on the changes a bit, I left a few comments

File: xdsl/transforms/riscv_register_allocation.py [hunk 7]
```diff
@@ -1,19 +1,80 @@
...
+                for op in block.walk():
+                    if not isinstance(op, RISCVOp):
+                        # Don't perform register allocations on non-RISCV-ops
+                        continue
+
+                    for result in op.results:
+                        assert isinstance(result.typ, RegisterType)
+                        if result.typ.data.name is None:
+                            # If we run out of real registers, allocate a j register
+                            if block_registers == list():
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/transforms/riscv_register_allocation.py:60]: I'm more a fan of explicitly testing `len(...) == 0`

[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/transforms/riscv_register_allocation.py:60]: Ah, actually PEP8 recommends this:
```suggestion
                            if not block_registers:
```
https://peps.python.org/pep-0008/

This makes sense as testing for empty is much more efficient than getting length in general. We should probably go with that, I'd say

File: xdsl/transforms/riscv_register_allocation.py [hunk 8]
```diff
@@ -57,16 +62,16 @@ def allocate_registers(self, module: ModuleOp) -> None:
                         assert isinstance(result.typ, RegisterType)
                         if result.typ.data.name is None:
                             # If we run out of real registers, allocate a j register
-                            if block_registers == list():
+                            if len(block_registers) == 0:
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:65]:
```suggestion
                            if not block_registers:
```

File: xdsl/transforms/riscv_register_allocation.py [hunk 9]
```diff
@@ -1,19 +1,85 @@
...
 
     def __init__(self) -> None:
         self.idx = 0
 
+        """
+        Since we've got neither right now a handling of a consistent ABI nor of a calling convention,
+        let's just assume that we have all the registers available for our use except the one explicitly reserved by the default riscv ABI.
+        """
+
+        self.available_registers = OrderedDict(Register.ABI_INDEX_BY_NAME)
```
Comment (superlopuh) [inline xdsl/transforms/riscv_register_allocation.py:37]: Looking at it more in detail, it seems like I gave a bad recommendation here. Only `reserved_registers` should be a set, the others were better the way you first wrote them, as lists.

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/transforms/riscv_register_allocation.py (hunk 9 above)
Comment (superlopuh) [reply xdsl/transforms/riscv_register_allocation.py:37]: My b

**Review by superlopuh — APPROVED** [review]
thank you!

File: tests/dialects/test_riscv.py [hunk 10]
```diff
@@ -89,6 +90,42 @@ def test_comment_op():
...
+
+    label_str = "mylabel"
+    label_op = riscv.LabelOp(f"{label_str}", region=label_region)
+
+    assert label_op.label.data == label_str
+
+    code = riscv_code(ModuleOp([label_op]))
+    assert code == f"{label_str}:\n    add a0, a1, a2\n"
+
+
```
Comment (AntonLydike) [inline tests/dialects/test_riscv.py:128]: There are still changes from the label op in this PR, or am I mistaken?

**Review by AntonLydike — APPROVED** [review]
Looks a lot better already, I think you still have some label PR changes in here. I'd suggest either rebasing or merging main into this branch so it contains only your changes.

File: tests/dialects/test_riscv.py (hunk 10 above)
Comment (compor) [reply tests/dialects/test_riscv.py:128]: yeah, I'd suggest rebasing on the `main` branch since that PR is now merged.

**Review by nazavode — APPROVED** [review]
